File: programmers/Lv3.43105_정수삼각형_dp.cpp
#include <bits/stdc++.h>

using namespace std;

int dp[500][500]={0};
//dfs풀이

int solution(vector<vector<int>> triangle) {
    int answer = 0;
    int height = triangle.size();
    dp[0][0] = triangle[0][0];

    for(int i=1; i<triangle.size(); i++){
        for(int j=0; j<triangle[i].size(); j++){
            if(j==0){
                dp[i][j] = dp[i-1][j] +triangle[i][j];
            }
            else if(j==triangle[i].size()-1){
                dp[i][j] = dp[i-1][j-1] + triangle[i][j];
            }
            else{
                dp[i][j] = max(dp[i-1][j-1],dp[i-1][j]) + triangle[i][j];
            }
            answer = max(answer, dp[i][j]);
        }
    }

    // for(int i=0; i<triangle[height-1].size(); i++){
    //     answer = max(answer, dp[height-1][i]);
    // }
    return answer;
}

int main(){

    vector<vector<int>> triangle = {{7}, {3, 8}, {8, 1, 0}, {2, 7, 4, 4}, {4, 5, 2, 6, 5}};
    
    cout << solution(triangle);

    return 0;
}

File: programmers/Lv3.43105_정수삼각형_dp.py
def solution(triangle):
    answer = 0
    height = len(triangle)
    dp = [[0]*50 for i in range(500)]
    
    dp[0][0] = triangle[0][0]
    
    for i in range(1,height):
        for j in range(len(triangle[i])):
            if j==0:
                dp[i][j] = dp[i-1][j]+triangle[i][j]
            elif i==j:
                dp[i][j] = dp[i-1][j-1]+triangle[i][j]
            else:
                dp[i][j] = max(dp[i-1][j-1],dp[i-1][j])+triangle[i][j]

    answer = max(dp[height-1])

    return answer

triangle = [[7], [3, 8], [8, 1, 0], [2, 7, 4, 4], [4, 5, 2, 6, 5]]

result = 30

print(solution(triangle))